### PR TITLE
i-bem__dom: update _elemCache on findElem

### LINKED ISSUE #583: i-bem, i-bem-dom: `findElem` should update elem's cache
Can `findElem` method rewrite `elem`'s internal cache if called with the same parameters?

### PR DESCRIPTION
**Pull request by qfox**
Fixes #583

### CODE REVIEW

File: common.blocks/i-bem/__dom/i-bem__dom.js [hunk 1]
```diff
@@ -723,13 +723,21 @@ var DOM = BEM.decl('i-bem__dom',/** @lends BEMDOM.prototype */{
...
             selector = '.' +
-                names.split(' ').map(function(name) {
+                names.map(function(name) {
                     return _self.buildClass(name, modName, modVal);
                 }).join(',.'),
             res = findDomElem(ctx, selector);
 
+        (ctx === this.domElem) && names.forEach(function(name) {
+                var key = _self.buildClass(name, modName, modVal),
+                    elem = this._elemCache[key] = res.filter('.' + key);
```
Comment (veged) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: In most cases we find a single elem (`names.length === 1`). So we can avoid unnecessary `res.filter`:

``` js
var isSingleName = names.length === 1;
...
elem = this._elemCache[key] = isSingleName? res : res.filter('.' + key);
```

Comment (qfox) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: Sounds reasonable, thanks!

Comment (qfox) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: Is there difference between `name + buildModPostfix(modName, modVal)` and `_self.buildClass(name, modName, modVal)` ?

Comment (veged) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: definitely — `name + buildModPostfix(modName, modVal)` more like violation of encapsulation of building full class

**Comment by qfox** [comment]
What's with coverage?

**Comment by narqo** [comment]
(just in case...): you can investigate the coverage decreasing [directly on Coveralls](https://coveralls.io/builds/1833468)

**Comment by qfox** [comment]
I didn't see anything about `common.blocks/i-bem/__dom/i-bem__dom.js` but see `common.blocks/i-bem/i-bem.vanilla.js` change.

**Comment by narqo** [comment]
Oh, sh.., looks like we have a fail positive here. Take a look at the build logs in Travis: there are few failed specs.

/cc @tadatuta @andrewblond We got a critical problem with enb-bem-specs here!

**Comment by qfox** [comment]
Fixed bug.
Made a little refactoring.
Squashed commits.

Need review ;-)

https://github.com/enb-bem/enb-bem-specs/issues/21

File: common.blocks/i-bem/__dom/i-bem__dom.js [hunk 2]
```diff
@@ -736,13 +736,23 @@ DOM = BEM.decl('i-bem__dom',/** @lends BEMDOM.prototype */{
...
         }
 
+        names = names.split(' ');
+
         var _self = this.__self,
-            selector = '.' +
-                names.split(' ').map(function(name) {
-                    return _self.buildClass(name, modName, modVal);
-                }).join(',.'),
+            keys = names.map(function (name) {
```
Comment (narqo) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: code style problem here

Comment (qfox) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: Sorry, fixed. How it doesn't caught on pre-commit?..

**Comment by blond** [comment]
> Oh, sh.., looks like we have a fail positive here. Take a look at the build logs in Travis: there are few failed specs.

@zxqfox, @narqo https://github.com/bem/bem-core/issues/908

File: common.blocks/i-bem/__dom/i-bem__dom.js [hunk 3]
```diff
@@ -736,13 +736,23 @@ DOM = BEM.decl('i-bem__dom',/** @lends BEMDOM.prototype */{
...
-                }).join(',.'),
+            keys = names.map(function(name) {
+                return _self.buildClass(name, modName, modVal);
+            }),
+            isSingleName = keys.length === 1,
+            selector = '.' + (isSingleName? keys[0] : keys.join(',.')),
             res = findDomElem(ctx, selector);
 
+        // caching results if possible
+        (ctx === this.domElem) && keys.forEach(function(key, i) {
```
Comment (narqo) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: As far as I remember, it's not save to compare jQuery-nodes in this way. Check this:

``` js
var link = $('.link:eq(0)');
link === $(link);   // ☚ false
```

Can you compare underlying DOM-nodes instead?

File: common.blocks/i-bem/__dom/i-bem__dom.js [hunk 4]
```diff
@@ -736,13 +736,23 @@ DOM = BEM.decl('i-bem__dom',/** @lends BEMDOM.prototype */{
...
+                var elem = this._elemCache[key] = isSingleName? res : res.filter('.' + key);
+                delete elem.prevObject; // drop link to original set after filtering
```
Comment (dfilatov) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: What is `prevObject`? I can't find any usage of it.

Comment (narqo) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: Can you point me, where this `prevObject` key came from?

File: common.blocks/i-bem/__dom/i-bem__dom.js [hunk 5]
```diff
@@ -768,15 +778,8 @@ DOM = BEM.decl('i-bem__dom',/** @lends BEMDOM.prototype */{
...
-        var key = name + buildModPostfix(modName, modVal),
-            res;
-
-        if(!(res = this._elemCache[key])) {
-            res = this._elemCache[key] = this.findElem(name, modName, modVal);
-            res.__bemElemName = name;
-        }
-
-        return res;
+        var key = this.__self.buildClass(name, modName, modVal);
```
Comment (narqo) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: Looks like this `key` variable can be inlined now?

Comment (zxqfox) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: `res.filter` for sure.

File: common.blocks/i-bem/__dom/i-bem__dom.js (hunk 4 above)
Comment (qfox) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: From `res.filter` for sure.

File: common.blocks/i-bem/__dom/i-bem__dom.js (hunk 5 above)
Comment (qfox) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: Yes, if you want.

File: common.blocks/i-bem/__dom/i-bem__dom.js (hunk 3 above)
Comment (qfox) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: You're right. Need to fix that.

File: common.blocks/i-bem/__dom/i-bem__dom.js (hunk 5 above)
Comment (qfox) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: Inlined.

File: common.blocks/i-bem/__dom/i-bem__dom.js (hunk 4 above)
Comment (qfox) [inline common.blocks/i-bem/__dom/i-bem__dom.js]: Fixed, added test. Check please.

There is `.is` function but I'm not sure about we should use it. It's pretty slow in compare with raw equal (but still much more readable): http://jsperf.com/jquery-objects-comparison/9

**Comment by dfilatov** [comment]
@zxqfox I'm still puzzled how `prevObject` is used.

**Comment by qfox** [comment]
@dfilatov Idea was to remove `res` (placed in `prevObject` key) to reduce links count and make possible to collect garbage. I can drop it if you don't like it.

**Comment by dfilatov** [comment]
I open i-bem__dom.js with your changes and try to find substring "prevObject".
There's only occurrence:
`delete elem.prevObject;`

**Comment by qfox** [comment]
It comes from here: https://github.com/jquery/jquery/blob/2380028ec4a6a77401b867a51de26a3cb8e8d311/src/core.js#L70

**Comment by dfilatov** [comment]
Oh, my god. Let's not to use jquery internals.

**Comment by qfox** [comment]
Okay, but `res` will never be collected ;-)

**Comment by dfilatov** [comment]
Why?

**Comment by qfox** [comment]
Because `res.filter` create new object for each `name`/`key` and `res` is stored in `prevObject` property in each new object. Nvm.

**Comment by dfilatov** [comment]
If I lose link to `res`, they will be collected, won't they?

**Comment by qfox** [comment]
That's why I tried to delete it. in other way it will be stored in `prevObject` property. Where I'm wrong? I can't get ;-)
As an opposite there is another idea to make empty chain (`$()`) and push there filtered elements. But is it worth?

**Pull request merged**